src/main/java/com/iota/iri/TransactionValidator.java

```diff
@@ -167,51 +170,50 @@ public void addSolidTransaction(Hash hash) {
-                while(cascadeIterator.hasNext() && !shuttingDown.get()) {
-                    try {
-                        Hash hash = cascadeIterator.next();
-                        TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hash);
-                        Set<Hash> approvers = transaction.getApprovers(tangle).getHashes();
-                        for(Hash h: approvers) {
-                            TransactionViewModel tx = TransactionViewModel.fromHash(tangle, h);
-                            if(quietQuickSetSolid(tx)) {
-                                tx.update(tangle, "solid");
-                            } else {
-                                if (transaction.isSolid()) {
-                                    addSolidTransaction(hash);
-                                }
-                            }
-                        }
-                    } catch (Exception e) {
-                        log.error("Some error", e);
-                        // TODO: Do something, maybe, or do nothing.
-                    }
-                }
-                synchronized (cascadeSync) {
-                    if (useFirst.get()) {
-                        newSolidTransactionsTwo.clear();
-                    } else {
-                        newSolidTransactionsOne.clear();
+                propagateSolidTransactions();
+            }
+        };
+    }
+
+    void propagateSolidTransactions() {
+        Set<Hash> newSolidHashes = new HashSet<>();
+        useFirst.set(!useFirst.get());
+        synchronized (cascadeSync) {
+            if (useFirst.get()) {
+                newSolidHashes.addAll(newSolidTransactionsTwo);
+            } else {
+                newSolidHashes.addAll(newSolidTransactionsOne);
+            }
+        }
+        Iterator<Hash> cascadeIterator = newSolidHashes.iterator();
+        while(cascadeIterator.hasNext() && !shuttingDown.get()) {
+            try {
+                Hash hash = cascadeIterator.next();
+                TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hash);
+                Set<Hash> approvers = transaction.getApprovers(tangle).getHashes();
+                for(Hash h: approvers) {
+                    TransactionViewModel tx = TransactionViewModel.fromHash(tangle, h);
+                    if(quietQuickSetSolid(tx)) {
+                        tx.update(tangle, "solid");
+                        addSolidTransaction(h);
                     }
                 }
-                try {
-                    Thread.sleep(500);
-                } catch (InterruptedException e) {
-                    e.printStackTrace();
-                }
+            } catch (Exception e) {
+                log.error("Error while propagating solidity upwards", e);
             }
-        };
+        }
+        synchronized (cascadeSync) {
+            if (useFirst.get()) {
+                newSolidTransactionsTwo.clear();
```

Comment:
Why do you not clear these directly after `addAll`?
This means that you'll potentially discard transactions that were added after your `addAll` above

Reply:
I didn't do it because I wanted to change as little as possible from the code.
Since you claim this is also a bug (and I think you are correct) I will fix it now.